app/views/automated_tests/student_interface.html.erb

```diff
@@ -4,15 +4,15 @@
 <%= javascript_include_tag 'result', nonce: true %>
 <%= javascript_include_tag 'Results/main', nonce: true %>
 
-<% if @assignment.enable_student_tests && !@grouping.nil? %>
+<% if @assignment.enable_student_tests && @grouping.present? %>
   <p>
     <strong><%= Assignment.human_attribute_name(:token_start_date) %>:</strong>
     <%= l(@assignment.token_start_date) %>
   </p>
 
   <p>
     <strong><%= Assignment.human_attribute_name(:token_end_date) %>:</strong>
-    <%= l(@assignment.token_end_date) %>
+    <%= @assignment.token_end_date.present? ? l(@assignment.token_end_date) : l(@assignment.due_date) %>
```

Comment:
This is on the right track but not exactly correct. Look at the `StudentPolicy#run_tests?` method, as that is what actually controls whether tests can be run or not. Look at the case for when `token_end_date.present` is false and go from there.